### PR TITLE
Enable -fdefer-type-errors on start-up

### PR DESCRIPTION
**Pull request by chrisdone**
See https://github.com/commercialhaskell/intero/issues/476#issuecomment-348467874 for detailed explanation.

I'm going to spend the day on this branch; if it doesn't cause any regressions in behavior (or bad performance), I'll merge to master.

If you want to test this out now, just run `killall intero` and start the processes up again.

### CODE REVIEW

**Comment by chrisdone** [comment]
@purcell Are you aware of any performance numbers showing that this is not usable? I'm going to test it on a few real world big files today and report if it's a bad experience.

**Comment by chrisdone** [comment]
Loading `Types.hs` from my Duet project (654 lines of types and derivings):

|Normal|With deferred type errors|
|:---|:--|
|13.658 seconds| 13.329 seconds|

No actual observed difference here.

**Comment by chrisdone** [comment]
From `Server.hs` from a client project, 

|Situation|Normal|With deferred type errors|
|:--|:--|:--|
|Type error|0.354 seconds|0.813 seconds|
|No type error|0.477 seconds|0.515 seconds|

So the numbers are slower when there's more code going on. Flagging up a type error is twice as slow, and flagging that the code is OK takes an extra 100ms.

**Comment by chrisdone** [comment]
It seems to vary based on caching or garbage collection inside the GHC API, because it goes as low as 

    0.643 seconds

So, really, roughly twice as slow at picking up type errors. You basically pay the price of doing a full type check every time you request a `:load`, whether it's a type error or not.

**Comment by purcell** [comment]
I feel like it's probably workable. It's a little slower for me on some files which were often annoyingly slow to flycheck anyway.

File: elisp/intero.el [hunk 1]
```diff
@@ -737,8 +738,10 @@ CHECKER and BUFFER are added to each item parsed from STRING."
                (msg (match-string 3)) ;; Replace gross bullet points.
                (type (cond ((string-match "^Warning:" msg)
                             (setq msg (replace-regexp-in-string "^Warning: *" "" msg))
-                            (if (string-match "^\\[-Wdeferred-type-errors\\]" msg)
-                                'error
+                            (if (or (string-match "^\\[-Wdeferred-type-errors\\]" msg)
```
Comment (purcell) [inline elisp/intero.el]: You can just use `string-prefix-p` here.

**Comment by chrisdone** [comment]
In the interest of always "working" (or, working more often), I think I'll merge. 

The speed hit does bother me a bit, but we can always be cleverer later to optimize the speed.

**Comment by purcell** [comment]
After using this for a day or two, I reckon this has actually made the common case of flycheck error reporting significantly worse: if you refer to an unimported function, for example, then deferred errors will be reported in the code which calls it, but the root problem itself will not be detected. I feel like reverting this might be the best course of action in the first time.

**Comment by chrisdone** [comment]
@purcell I think I pushed a change to fix that - can you try latest master? 

I noticed the same problem; out of scope identifiers become typed holes with deferred type errors on.

Can you still reproduce your problem?

**Comment by purcell** [comment]
Yeah, I _think_ it's better now. Will plug away and report back. :-)

**Comment by chrisdone** [comment]
@purcell Cool!

**Pull request merged**